teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+---
+status: proposed
+title: Decouple Task Composition from Scheduling
+creation-date: '2021-01-22'
+last-updated: '2021-02-08'
+authors:
+- '@bobcatfish'
+---
+
+# TEP-0044: Decouple Task Composition from Scheduling
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Use Cases (optional)](#use-cases-optional)
+  - [Overlap with TEP-0046](#overlap-with-tep-0046)
+- [Requirements](#requirements)
+- [References (optional)](#references-optional)
+  - [PipelineResources](#pipelineresources)
+- [Design Details](#design-details)
+- [Alternatives](#alternatives)
+
+## Summary
+
+This TEP addresses how the current concept of a Task embodies both a reusable unit of logic but is also a unit used to
+schedule execution.
+
+## Motivation
+
+* Tasks are a reusable unit combining one or more steps (container images) together and defining an interface to those
+  steps via params, workspaces and results
+* Tasks can be combined in a Pipeline
+* When a Pipeline is invoked via a PipelineRun, each Task is executed as a separate pod.
+
+This means that choices made around Task design (e.g. creating a Task that encapsulates a git clone and a separate Task
+to run go unit tests) directly impact the performance and overhead involved in executing the Tasks. For example
+if the git clone task wants to share data with the unit test task, beyond a simple
+[result](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#using-results), it needs to 
+you'll need to provision a PVC or do some other similar, cloud specific storage,
+to [make a volume available](https://github.com/tektoncd/pipeline/blob/master/docs/workspaces.md#specifying-volumesources-in-workspaces)
+that can be shared between them, and running the second Task will be delayed by the overhead of scheduling a second pod.
+
+### Goals
+
+- Make it possible to combine Tasks together so that you can run multiple
+  Tasks together and have control over the scheduling overhead (i.e. pods and volumes required) at authoring time
+  in a way that can be reused (e.g. in a Pipeline)
+- Add some of [the features we don't have without PipelineResources](https://docs.google.com/document/d/1KpVyWi-etX00J3hIz_9HlbaNNEyuzP6S986Wjhl3ZnA/edit#)
+  to Tekton Pipelines (without requiring use of PipelineResources), specifically **Task adapters/specialization**
```

Comment:
What is "Task adapters/specialization" in the TEP ? (I feel the term is used there without prior definition or example)

Reply:
this is trying to refer to the content of the linked doc:

![image](https://user-images.githubusercontent.com/432502/110861345-007a9d00-828c-11eb-919e-c301045d609f.png)

ill reword this and try to make it more clear